lib/rules/spinner-needs-labelling.js

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+"use strict";
+
+const { hasNonEmptyProp } = require("../util/hasNonEmptyProp");
+const elementType = require("jsx-ast-utils").elementType;
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        // possible error messages for the rule
+        messages: {
+            noUnlabelledSpinner: "Accessibility: Spinner must have aria-label and aria-live"
+        },
+        // "problem" means the rule is identifying code that either will cause an error or may cause a confusing behavior: https://eslint.org/docs/latest/developer-guide/working-with-rules
+        type: "problem",
+        // docs for the rule
+        docs: {
+            description: "Accessibility: Spinner must have aria-label and aria-live",
+            recommended: true,
+            url: "https://www.w3.org/TR/html-aria/" // URL to the documentation page for this rule
+        },
+        schema: []
+    },
+    // create (function) returns an object with methods that ESLint calls to “visit” nodes while traversing the abstract syntax tree
+    create(context) {
+        return {
+            // visitor functions for different types of nodes
+            JSXOpeningElement(node) {
+                // if it is not a Spinner, return
+                if (elementType(node) !== "Spinner") {
+                    return;
+                }
+
+                if (hasNonEmptyProp(node.attributes, "aria-label") && hasNonEmptyProp(node.attributes, "aria-live")) {
```

Comment:
We might also want to consider checking if the aria-busy attribute is present. 

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-busy

https://dockyard.com/blog/2020/03/02/accessible-loading-indicatorswith-no-extra-elements

Reply:
Got it.Curious, when I inspected spinner on this page: https://react.fluentui.dev/?path=/docs/components-spinner--docs#size , it doesn't seems to have an aria-busy attribute. What happens in this case?

Reply:
Under the hood, it's a progress bar: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/progressbar_role It looks like aria-labelledby is also a valid approach to labelling.

Reply:
Using aria-live is a great way to make dynamic content updates accessible to screen reader users, but it functions differently from aria-busy. Here’s what would happen if you used aria-live on a spinner or loading indicator instead of (or alongside) aria-busy:

How aria-live Works:
aria-live is used to announce changes in the content of a specific region to assistive technologies. You can set aria-live to different levels:

aria-live="polite": Content is announced when the user is idle or not interacting with other elements, allowing less critical updates to be non-disruptive.
aria-live="assertive": Content is announced immediately, even if the user is interacting with something else, for high-priority updates.
What Happens If You Use Only aria-live on a Loading Indicator:
Content Updates are Announced: If you use aria-live="polite" or aria-live="assertive" on a spinner or loading message, screen readers will announce changes to the content when the indicator appears or when the content loads.

Example: If you have a loading spinner and set aria-live="polite", the screen reader will announce, "Loading..." when the spinner is rendered, and then announce "Content loaded" when the spinner is replaced by the loaded content.
No Busy State Information: aria-live doesn't indicate to users that the current region is in the process of being updated like aria-busy does. Without aria-busy, screen reader users won’t know that content is still being processed and may try to interact prematurely with content that is not yet fully loaded.

Impact: Users may start interacting with incomplete content without realizing that the page is still loading.

Multiple Announcements: Using aria-live without aria-busy could cause the screen reader to announce every update in real-time. This might be useful for some situations but could also result in repetitive or overwhelming notifications if the content updates frequently.

Example: If aria-live="assertive" is set on a spinner that continuously updates, the screen reader might announce "Loading..." over and over again, potentially overwhelming users.

Best Practices:
Use aria-busy for Loading Indicators: aria-busy is more appropriate for signaling that content is being updated (e.g., while loading). It helps assistive technologies notify users that a region is still loading or being updated, preventing them from interacting with incomplete content.
Use aria-live for Content Changes: aria-live is better suited for announcing updates or new content once it’s loaded (e.g., when loading is complete or content changes dynamically). For example, once the content loads, you can set aria-live="polite" on the newly loaded content to announce it.
Combining aria-busy and aria-live:
You can use both aria-busy and aria-live together to provide the best experience:

Use aria-busy="true" while content is loading to indicate that a region is still in progress.
Use aria-live="polite" to announce once the content has been loaded.
Example:

html
Copy code
<div aria-live="polite" aria-busy="true">
  Loading content, please wait...
</div>
<!-- When content is loaded -->
<div aria-live="polite" aria-busy="false">
  Content loaded successfully.
</div>
Conclusion:
aria-busy is critical for letting users know that content is still loading, preventing premature interaction.
aria-live helps announce changes to content once it's loaded, but doesn’t indicate a "loading" state like aria-busy.
Best practice: Use both aria-busy and aria-live together for spinners or loading indicators to ensure accessible and clear feedback for screen reader users.

Reply:
Damn, that is very neat use of aria-live and aria-busy. Thanks a lot for explaining it!